core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,222 @@
+use std::{any::Any, collections::HashMap, fmt};
+
+use futures::{future::BoxFuture, FutureExt};
+use tokio::{runtime::Runtime, sync::watch};
+
+pub use self::{context::NodeContext, stop_receiver::StopReceiver};
+use crate::{
+    resource::ResourceProvider,
+    task::{IntoZkSyncTask, TaskInitError, ZkSyncTask},
+};
+
+mod context;
+mod stop_receiver;
+
+type TaskConstructor =
+    Box<dyn FnOnce(NodeContext<'_>) -> Result<Box<dyn ZkSyncTask>, TaskInitError>>;
+
+/// "Manager" class of the node. Collects all the resources and tasks,
+/// then runs tasks until completion.
+///
+/// Initialization flow:
+/// - Node instance is created with access to the resource provider.
+/// - Task constructors are added to the node. At this step, tasks are not created yet.
+/// - Optionally, a healthcheck task constructor is also added.
+/// - Once the `run` method is invoked, node
+///   - attempts to create every task. If there are no tasks, or at least task
+///     constructor fails, the node will return an error.
+///   - initializes the healthcheck task if it's provided.
+///   - waits for any of the tasks to finish.
+///   - sends stop signal to all the tasks.
+///   - waits for the remaining tasks to finish.
+///   - calls `after_node_shutdown` hook for every task that has provided it.
+///   - returns the result of the task that has finished.
+pub struct ZkSyncNode {
+    /// Primary source of resources for tasks.
+    resource_provider: Box<dyn ResourceProvider>,
```

Comment:
IMO given that ResourceProvider is not the only source of resources now, why do we need it in the first place? Wouldn't it be more consistent to just have an add_resource method on ZkSyncNode to prepopulate resources if needed and otherwise just wait for tasks to provide them?

Reply:
I guess the main thing here is that ResourceProvider is lazy, it'll only provide resources if required.
Determining whether we need an `ObjectStore`, for example, would be hard -- we'll need to go through the list of tasks, somehow check whether any of them needs an `ObjectStore`, and then add it.

Reply:
if tasks are supposed to be able to add resources as well, then one will have to do this reasoning anyway (since tasks are not lazy), no?

Reply:
That's true, but then we'll have to ship config for dependency resources together with all the tasks.
OTOH this can be solved with "ConfigProvider", hm.

Reply:
Although not really -- we will have to make every task aware of every possible `ObjectStore` implementation, which is one more thing we want to avoid.

I think I'll stick with `ResourceProvider` for now, but maybe we'll brainstorm how to improve the system once the codebase is adapted to support the new framework.

Reply:
Well, task-wise, yes, one is expected to provide a correct set of tasks.
I see your point, but right now it feels like creating a correct set of tasks is simpler than creating a correct set of resources (based on the current state of [`initialize_components`](https://github.com/matter-labs/zksync-era/blob/main/core/lib/zksync_core/src/lib.rs#L272)).

If it'll turn out to be untrue, it should be pretty straightforward to remove `ResourceProvider` completely later.

Reply:
Regarding resource lazy instantiation, I believe it could be dealt with by lazy resources. IMO, it'd be relatively maintainable to define a "task" / set of "tasks" (`ResourceConfiguration` as per other line comment) that would fulfill the `ResourceProvider` role (i.e., provide resources based on a configuration). And we could still retain "statically known" (node config-based) set of resources. This would also allow making `StoredResource` crate-private (which, IMO, it should be).